functions/db/deliveryRequests/onUpdate.f.js

```diff
@@ -0,0 +1,115 @@
+const functions = require('firebase-functions');
+const enums = require('../../Enums.js');
+const utils = require('../../Utils.js');
+
+const nt = enums.NotificationType;
+
+/*
+ * Listeners based on changes to recurring pickup requests:
+ *
+ * LISTENER 1: handle RA responses for recurring pickup requests
+ * Cases:
+ *  A) a RA claimed:
+ *      Change: 'receivingAgency' changed from requested/pending to claimed
+ *      Action: send notifications to DGs
+ *  B) all RAs rejected:
+ *      Change: 'status' changed from PENDING to REJECTED_RA
+ *      Action: send notification back to DA
+ *  C) no RA responded in time:
+ *      TODO: to be handled by a different scheduled function
+ *
+ * Listener 2: handle DG responses for recurring pickup requests
+ * TODO
+ */
+exports = module.exports = functions.database
+    .ref('/delivery_requests/{umbrellaId}/{pushId}')
+    .onUpdate((change, context) => {
+        console.info('Recurring request changed: ' + context.params.pushId);
+
+        // once a request is not in pending status, it shouldn't be changed anymore
+        if (change.before.val().status !== enums.RequestStatus.PENDING) {
+            return Promise.reject(
+                new Error('Changes were made to a non-pending pickup request.'));
+        }
+
+        let requestSnap = change.after;
+        let request = requestSnap.val();
+
+        // get db refs. TODO: set up Admin SDK
+        const rootRef = change.after.ref.parent.parent.parent;
+        const accountsRef = rootRef.child('accounts');
+        const daRef = rootRef.child(`donating_agencies/${request.donatingAgency}`);
+
+        // ---------- Listener 1 triggers ----------
+        if (raClaimed(change)) {  // case A
+            return sendRequestToDGs(accountsRef, requestSnap);
+        }
+
+        if (raRejected(change)) {  // case B
+            console.info('Listener1: all RA rejected -> send notification to DA');
+            return utils.notifyRequestUpdate(
+                'DA', daRef, requestSnap.key, nt.RECURRING_PICKUP_REJECTED_RA);
+        }
+        // ---------- Listener 1 end ----------
+
+        // TODO: Listener 2
+
+        console.info('Nothing was triggered.');
+        return null;
+    });
+
+// ----------------------- Listener 1 funcs -----------------------
+// case A trigger
+function raClaimed(change) {
+    return !change.before.child('receivingAgency').hasChild('claimed') && 
+        change.after.child('receivingAgency').hasChild('claimed');
+}
+
+// case B trigger
```

Comment:
should be checking the length of the pending list, not the status because clients might update this at the same time and unreliably.